app/recipe/tests/test_recipe_api.py

```diff
@@ -0,0 +1,125 @@
+        'time_minutes': 10,
+        'price': 5.00,
+    }
+    defaults.update(params)
+
+    return Recipe.objects.create(**defaults)
+
+
+class PublicRecipeApiTests(TestCase):
+    """Test recipe management"""
+
+    def setUp(self):
+        self.client = APIClient()
+
+    def test_retrieve_recipes(self):
+        """Test getting a list of recipes"""
+        sample_recipe()
+        sample_recipe()
+
+        res = self.client.get(RECIPES_URL)
+
+        recipes = Recipe.objects.all().order_by('-id')
+        serializer = RecipeSerializer(recipes, many=True)
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        self.assertEqual(
+            res.data[0].get('title'), serializer.data[0].get('title')
+        )
+
+    def test_view_recipe_detail(self):
+        """Test viewing a single recipe detail"""
+        recipe = sample_recipe()
+        recipe.ingredients.add(sample_ingredient())
+
+        url = detail_url(recipe.id)
+        res = self.client.get(url)
+
+        serializer = RecipeSerializer(recipe, many=False)
+        self.assertEqual(res.data, serializer.data)
+
+    def test_create_basic_recipe(self):
+        """Test creating a new delicious recipe"""
+        payload = {
+            'title': 'Carrot cake muffins',
+            'time_minutes': 30,
+            'price': 4.00,
+        }
+        res = self.client.post(RECIPES_URL, payload)
+
+        self.assertEqual(res.status_code, status.HTTP_201_CREATED)
+        recipe = Recipe.objects.get(id=res.data['id'])
+        for key in payload.keys():
+            self.assertEqual(payload[key], getattr(recipe, key))
+
+    def test_partial_update_recipe(self):
+        """Test updating a recipe with patch"""
+        recipe = sample_recipe()
+        recipe.ingredients.add(sample_ingredient())
+        new_ingredient = sample_ingredient(name='Curry')
+
+        payload = {'title': 'Roasted tomato risotto', 'ingredients': [new_ingredient.id]}
+        url = detail_url(recipe.id)
+        self.client.patch(url, payload)
+
+        recipe.refresh_from_db()
+        self.assertEqual(recipe.title, payload['title'])
+        ingredients = recipe.ingredients.all()
+        self.assertEqual(len(ingredients), 1)
+        self.assertIn(new_ingredient, ingredients)
+
+    def test_full_update_recipe(self):
+        """Test updating a recipe with put"""
+        recipe = sample_recipe()
+        recipe.ingredients.add(sample_ingredient())
+
+        payload = {
+            'title': 'Margarita pizza',
+            'time_minutes': 15,
+            'price': 7.00
+        }
+        url = detail_url(recipe.id)
+        self.client.put(url, payload)
+
+        recipe.refresh_from_db()
+        self.assertEqual(recipe.title, payload['title'])
+        self.assertEqual(recipe.time_minutes, payload['time_minutes'])
+        self.assertEqual(recipe.price, payload['price'])
+        ingredients = recipe.ingredients.all()
+        self.assertEqual(len(ingredients), 0)
```

Comment:
Here'd be a classical case of premature improvement, but in many other cases using length would be slower.
You retrieve the full list of ingredients which in other cases could be a big table with many entries, just to count them later on. Better use 
```suggestion
        self.assertEqual(ingredients.count(), 0)
```

Or, even better in this particular case where you expect 0
```
self.assertFalse(ingredients.exists())
```

Reply:
Interesting, the Udemy course touched on this briefly as well giving both as options. I guess I picked the 'wrong one' in this case then :D